src/wasm-type.h

```diff
@@ -90,18 +106,18 @@ class Type {
   bool hasRef() { return hasPredicate<&Type::isRef>(); }
 
   constexpr uint64_t getID() const { return id; }
-  constexpr ValueType getSingle() const {
-    assert(!isMulti() && "Unexpected multivalue type");
-    return static_cast<ValueType>(id);
+  ID getSingle() const {
+    assert(!isTuple() && "Unexpected tuple type");
+    return static_cast<ID>(id);
```

Comment:
Hmm, `getSingle` is odd in that it includes `none` and `unreachable` while `isSingle` does not. Noticed this while trying to make it a `constexpr` again using `isSingle`, but then it excluded `none` and `unreachable` and would error.

In calling code, it's used to obtain the basic type most of the time (should rather be named `isBasic` there), but not always, for example some checks use it according to the implementation to check that something is not a tuple, like the type of an exported global. Please advise.

Reply:
I agree that this is kind of a weird conflation of separate concepts. For the majority of uses, including all `switch`es over types, you're right that `getBasic` (and a corresponding `isBasic`) would be a better name. I also agree that ultimately we will need both. Fixing this usage throughout the code base would be a good preliminary PR, although somewhat tedious.

Reply:
Revising the call sites is definitely a good thing to do since we'll have to do this anyway to support the new constructed types at the call sites of `getSingle` properly on a case-by-case basis. Support these makes it hard to do this in a preliminary PR, though. Not sure what's best.

Reply:
I think a PR just introducing `isBasic` and `getBasic` and fixing up `isSingle` and `getSingle` to not include `Type::none` or `Type::unreachable`, then replacing most uses of `isSingle` (`getSingle`) with `isBasic` (`getBasic`) would be good. I don't think that would have any dependency on the new `TypeDef`.

Reply:
Alright, will try :)

Reply:
This is harder than expected. Take for instance

```cpp
            switch (curr->operands[i]->type.getSingle()) {
              case Type::i32:
                curr->operands[i] = parent->builder.makeUnary(
                  ConvertSInt32ToFloat64, curr->operands[i]);
                break;
              case Type::f32:
                curr->operands[i] =
                  parent->builder.makeUnary(PromoteFloat32, curr->operands[i]);
                break;
              default: {
              } // f64, unreachable, etc., are all good
            }
```

This clearly wants single types since it deals with call operands, but `unreachable` and `none` are also ok. Looking at this from a purely functional perspective, it is fine to replace `getSingle` with `getBasic`, indeed it is even necessary to make it work, but it is semantically incorrect. Like, once new constructed types being `isSingle` are introduced, `getBasic` will explode. A circular dependency on this PR one could say :)

Reply:
To also have some good news, I made a new diagram to put into the header:

```cpp
  // Predicates
  //                         C
  //                         o
  //                     S   n
  //                 B   i   c
  //                 a   n   r
  //                 s   g   e
  //                 i   l   t
  //   Type          c   e   e
  // ╞═════════════╬═══╪═══╪═══╪═══════╡
  // │ none        ║ x │   │   │       │
  // │ unreachable ║ x │   │   │       │
  // ├─────────────╫───┼───┼───┤───────┤
  // │ i32         ║ x │ x │ x │ I     │ ┐ Number
  // │ i64         ║ x │ x │ x │ I     │ │  I_nteger
  // │ f32         ║ x │ x │ x │   F   │ │  F_loat
  // │ f64         ║ x │ x │ x │   F   │ │  V_ector
  // │ v128        ║ x │ x │ x │     V │ ┘
  // ├─────────────╫───┼───┼───┤───────┤
  // │ funcref     ║ x │ x │ x │ f     │ ┐ Ref
  // │ externref   ║ x │ x │ x │       │ │  f_unc
  // │ nullref     ║ x │ x │ x │       │ │
  // │ exnref      ║ x │ x │ x │       │ │
  // ├─────────────╫───┼───┼───┤───────┤ │
  // │ Signature   ║   │ x │ x │ f     │ │ ┐
  // │ Struct      ║   │ x │ x │       │ │ │ TODO (GC)
  // │ Array       ║   │ x │ x │       │ ┘ ┘
  // │ Multi       ║   │   │ x │       │
  // └─────────────╨───┴───┴───┴───────┘
```

Unfortunately it also illustrates the problem with `Signature`, `Struct` and `Array` being `isSingle`, with the overlap of `getBasic` and `getSingle` being asymmetric.

Reply:
> This is harder than expected. Take for instance
> 
> ```c++
>             switch (curr->operands[i]->type.getSingle()) {
>               case Type::i32:
>                 curr->operands[i] = parent->builder.makeUnary(
>                   ConvertSInt32ToFloat64, curr->operands[i]);
>                 break;
>               case Type::f32:
>                 curr->operands[i] =
>                   parent->builder.makeUnary(PromoteFloat32, curr->operands[i]);
>                 break;
>               default: {
>               } // f64, unreachable, etc., are all good
>             }
> ```
> 
> This clearly wants single types since it deals with call operands, but `unreachable` and `none` are also ok. Looking at this from a purely functional perspective, it is fine to replace `getSingle` with `getBasic`, indeed it is even necessary to make it work, but it is semantically incorrect. Like, once new constructed types being `isSingle` are introduced, `getBasic` will explode. A circular dependency on this PR one could say :)

I see what you mean. For example like this, I would change the `getSingle()` to a `getBasic()` and add the following two assertions before the whole block:

```c++
assert(curr->operands[i]->type.isSingle());
assert(curr->operands[i]->type.isBasic() && "TODO: handle constructed types");
```

We can worry about correctly resolving those `TODO: handle constructed types` in follow-up PRs.